### PR TITLE
[WIP] feat: JSON Deserializer/Denormalizer

### PR DESCRIPTION
**Pull request by lal12**
Implements (half of) #86 

Mostly done (I hope :D), but still have some issues with tests. Happy accepting advise.

### CODE REVIEW

File: src/serialize/jsonDeserializer.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    options: DenormalizerOptions = {}
+  ): Bom {
+    return this.#normalizerFactory.makeForBom()
+      .denormalize(data, options)
+  }
+
+  protected _deserialize (
+    str: string,
+    opts: DesserializerOptions = {}
+  ): Normalized.Bom {
```
Comment (jkowalleck) [inline src/serialize/jsonDeserializer.ts:55]: :x: 
`JSON.parse()` can return anything, it is not guaranteed to return `Normalized.Bom`
The return type is under control of the users, because they control the input string.
Therefore, the algorithm is not robust.

File: src/serialize/json/denormalize.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  abstract denormalize (data: TNormalized, options: DenormalizerOptions): TModel | undefined
+}
+
+export class BomDenormalizer extends BaseJsonDenormalizer<Models.Bom> {
+  denormalize (data: Normalized.Bom, options: DenormalizerOptions): Models.Bom {
+    const spec = SpecVersionDict[data.specVersion as SpecVersion]
+    if (spec === undefined) {
+      throw new UnsupportedFormatError(`Spec version ${data.specVersion} is not supported.`)
+    }
+    this.factory.spec = spec
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:117]: https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506/files#diff-facac0ec4128527fecf05b9814c75d81c08dfe23a5bc7c9f26611f6e224b3c51R112-R117

this should be part of the tool that does initial analysis.
reading which spec version is applied should be done before the factory is initialized.

File: src/serialize/json/denormalize.ts [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    if (!this.factory.spec.supportsFormat('json')) {
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:118]: there are enums for the format. please use it.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
Thanks for your work,
I know there is still a lot to do, take your time.

File: src/serialize/jsonDeserializer.ts (hunk 1 above)
Comment (lal12) [reply src/serialize/jsonDeserializer.ts:55]: What do you propose?
Is there a JSON schema somewhere to validate against?

Comment (jkowalleck) [reply src/serialize/jsonDeserializer.ts:55]: Sure, schema file is included in the library. But there is no JSON schema validator included.

https://github.com/CycloneDX/cyclonedx-javascript-library/issues/86#issuecomment-1442152762

**Comment by lal12** [comment]
Worked on the denormalizer. Probably less flexible than you wished, but silently ignoring known properties if their type isn't right does not seem good behavior to me either.

No tests yet, but want feedback (@jkowalleck) on the general approach.

File: src/serialize/json/denormalize.ts [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+  Repository: new(arr: VT[]) => RT
+): RT | undefined {
+  assertArrayOrNull(arr, path)
+  return (arr != null)
+    ? new Repository(arr.map((item: any, index) => denormalizer.denormalize(item, options, [...path, index])))
+    : undefined
+}
+
+export class Factory {
+  #spec?: Spec
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:95]: :question: the fact that this is not `readonly` makes me wonder what would happen,
if I used this class in a parallel thread? 
I mean, JavaScript is all about async and event based and in general parallel.

That would happen if i used the same object with two different deserializes on two different BOMs?
They would change the `#spec` sight on use, which might result in unpredictable results. 
Please have this fixed. See https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506/files#r1115847206

File: src/serialize/json/denormalize.ts [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+
+  get spec (): Spec | undefined {
+    return this.#spec
+  }
+
+  set spec (spec: Spec | undefined) {
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:101]: no setter! 
see my remarks at the `#spec` definition.

File: src/serialize/json/denormalize.ts (hunk 2 above)
Comment (jkowalleck) [reply src/serialize/json/denormalize.ts:185]: you might wonder how the factory should be injectable, while it is not fully initialized? 
possible solution: Inject a factory method, or the class type, instead of an object o the class.

**Review by jkowalleck — CHANGES_REQUESTED** [review]
reviewed as [requested](https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506#issuecomment-1451935970)

remarks: 
- :+1: I like the implementation of type assertions. good job.
- :x: there is an option in the code, to [change](https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506#discussion_r1123662661) the `#spec` of a denormalizer `Factory` at runtime while this factory is in use. This side effect must be removed to prevent inconsistency.
- :-1: : I don't like the unforgiving behavior when data is not as expected. You could have an injectable optional `warningFunc: (msg: string) => void` that is called each time a value is unexpected or wrong, right before the value is ignored. It is important for the final result, to be able to generate a BOM model from partially invalid data. :100: This feature could still be added later on.

**Comment by jkowalleck** [comment]
re: https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506#pullrequestreview-1322565538

@lal12 i have it more thought and i thing the type checkers you were implementing are technically correct, but do not help TypeScript with the type assertions.
Could you implement them as proper type assertions or type predicates? 

read https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates
feature intro: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

example: 
```TypeScript
function isString (v: any): v is string 
{
  return typeof v === 'string'
}

function assertIsString (v: any): assert v is string 
{
  if ( typeof v !== 'string' ) { 
    throw new AssertionError() 
  }
}

example code of a dynamic/templated type predicate: ``
https://github.com/CycloneDX/cyclonedx-javascript-library/blob/9367f5d13c42abd8d48e05c2e83f2d840da70340/src/_helpers/notUndefined.ts#L20-L24

**Comment by lal12** [comment]
@jkowalleck 
I already (partially) do `function assertNonEmptyStr (value: any, path: PathType): asserts value is string {` (here https://github.com/lal12/cyclonedx-javascript-library/blob/3e014d71c84d129e0384953f97493e8814771231/src/serialize/json/denormalize.ts#L45)

I still had it on my list to look into `assertTypes (value: any, expected: VarType[], path: PathType): void`, whether some kind of generic mapped assertion is possible (probably). 

But I will refactor/change that anyway to get it working with `warningFunc: (msg: string) => void`.

**Comment by jkowalleck** [comment]
@lal12 
General question: why do you compare values against null - like `value != null`? 
As far as I know, none of the CycloneDX JSON schemas has any nullable value.
They have optional, which would be `undefined`, not `null` -- so you should compare `value !== undefined`.

**Comment by lal12** [comment]
@jkowalleck something I've noticed while writing tests is that normalizer will look for BomRef by reference to the object in https://github.com/CycloneDX/cyclonedx-javascript-library/blob/9367f5d13c42abd8d48e05c2e83f2d840da70340/src/serialize/json/normalize.ts#L501 instead of by it's string value. Is this intended? 

Made tests working, not ideal, but comparing boms fails due to URLs and `loadNormalizeResult` drops undefined, so I have to renormalize and stringify. 

Waiting for your review

File: src/serialize/jsonDeserializer.ts (hunk 1 above)
Comment (lal12) [reply src/serialize/jsonDeserializer.ts:53]: closing, is obsolete

**Comment by jkowalleck** [comment]
> @jkowalleck something I've noticed while writing tests is that normalizer will look for BomRef by reference to the object in
> 
> https://github.com/CycloneDX/cyclonedx-javascript-library/blob/9367f5d13c42abd8d48e05c2e83f2d840da70340/src/serialize/json/normalize.ts#L501
> instead of by it's string value. Is this intended?
yes. is intended. 
read https://github.com/CycloneDX/cyclonedx-php-library/blob/master/docs/dev/decisions/BomDependencyDataModel.md
Background: BomRef's [must be unique.](https://github.com/CycloneDX/specification/blob/ccbf7b5781ef534cd62616e3c4221004c7c82a66/schema/bom-1.4.xsd#L2402-L2405) Therefore, the values are not final and will be discriminated - see https://github.com/CycloneDX/cyclonedx-php-library/blob/5f475ce38750745a45f920664ec1acaf0aa02b1a/src/Core/Serialization/BaseSerializer.php#L73-L74


> web build complains that there is no URL polyfill, should I add one?
will check this later.

> Waiting for your review
will review soon

**Comment by lal12** [comment]
I import `URL` which the web build doesn't like, without further config. Just gotta remove it.

**Comment by jkowalleck** [comment]
> I import `import {URL} from 'url';` which the web build doesn't like, without further config. Just gotta remove it.

just remove this import. URL is a global available class in JS.

File: src/serialize/json/denormalize.ts [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+
+import { PackageURL } from 'packageurl-js'
+
+import { AttachmentEncoding, ComponentScope, ComponentType, ExternalReferenceType, HashAlgorithm } from '../../enums'
+import * as Models from '../../models'
+import type { Protocol, Protocol as Spec } from '../../spec'
+import { Format, SpecVersionDict, UnsupportedFormatError } from '../../spec'
+import type { JSONDenormalizerOptions, JSONDenormalizerWarning, PathType } from '../types'
+
+interface JSONDenormalizerContext {
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:28]: why is this needed? why are these three values encapsulated into a record? 
i found rare usage - and none was justified. all usage i found was artificially complex.

File: src/serialize/json/denormalize.ts [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+  }
+
+  abstract denormalize (data: TNormalized, ctx: JSONDenormalizerContext, path: PathType): TModel
+}
+
+export class BomDenormalizer extends BaseJsonDenormalizer<Models.Bom> {
+  denormalize (data: Record<string, unknown>, { options }: { options: JSONDenormalizerOptions }, path: PathType): Models.Bom {
+    assertEnum(data.bomFormat, ['CycloneDX'], [...path, 'bomFormat'])
+    assertEnum(data.specVersion, Object.keys(SpecVersionDict), [...path, 'specVersion'])
+    const spec = SpecVersionDict[data.specVersion as keyof typeof SpecVersionDict] as Protocol
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:275]: why is this `as keyof typeof SpecVersionDict` needed? the `assertEnum` should have done this already. if not, then it is done wrong.

i dont see a real need for this complex usage. please simplify
```ts
const spec = SpecVersionDict[data.specVersion as Version]
if (!spec) { throw Error('something') } 
```

File: src/serialize/json/denormalize.ts [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+    if (!spec.supportsFormat(Format.JSON)) {
```
Comment (jkowalleck) [inline src/serialize/json/denormalize.ts:276]: :x: NO!
the (de)normalizer does not decide whether to run or not - it just does the job.
the (de)serializer might decide to not run because of constraints.

**Review by jkowalleck — CHANGES_REQUESTED** [review]
did not review all of it, stopped at a point.

Generally I see a big wrong concept here, which i will try to explain
- you are deserializing the JOSN to some data, and force the data to be denormalized. while the denormalization is already ongoing, it is decided to stop all of this, because something seams odd.
  THIS MUST NOT BE.
- correct would be: 
  1. Deserialize the JSON to some data
  2. check if this data is an object, that has the property "bomFormat" set to "CycloneDX"  - if not STOP/THROW
  3. check if this data is an object, that has the property "specVersion" set to "one value that is known"
  4. fetch the Spec from `SpecVersionDict` based on the "specVersion" properrty. If it is not existing: Exit/Throw. If the Spec does not support JSON: Exit/Throw
  5. feed the data model into the denormalizer.

very basic (pseudo) example code to show the process: 
```ts
Deserializer {
  readonly #df
  constructor(df: DenormalizerFactory) {
    this.df=df
  }
  deserialize(bom: string) {
    const data = JSON.parse(bom)
    if (typeof data !== 'object') { throw RangeError('no object') }
    if (data.bomFormat !== 'CycloneDX') { throw RangeError('wrong format') }
    const spec = SpecVersionDict[data.specVersion]
    if (! spec?.supportsFormat('JSON')) { throw SpecVersionDict('something') }
    this.#df.makeForBom().denormalize(data)
  }
}
// i dont see a need to inject the spec into the normalizer.
```

**Comment by jkowalleck** [comment]
FYI: i will be working on #620 soon.
